File: python/3.longest-substring-without-repeating-characters.py
#
# @lc app=leetcode.cn id=3 lang=python3
#
# [3] 无重复字符的最长子串
#
# https://leetcode-cn.com/problems/longest-substring-without-repeating-characters/description/
#
# algorithms
# Medium (35.87%)
# Total Accepted:    734.5K
# Total Submissions: 2M
# Testcase Example:  '"abcabcbb"'
#
# 给定一个字符串，请你找出其中不含有重复字符的 最长子串 的长度。
# 
# 示例 1:
# 
# 输入: "abcabcbb"
# 输出: 3 
# 解释: 因为无重复字符的最长子串是 "abc"，所以其长度为 3。
# 
# 
# 示例 2:
# 
# 输入: "bbbbb"
# 输出: 1
# 解释: 因为无重复字符的最长子串是 "b"，所以其长度为 1。
# 
# 
# 示例 3:
# 
# 输入: "pwwkew"
# 输出: 3
# 解释: 因为无重复字符的最长子串是 "wke"，所以其长度为 3。
# 请注意，你的答案必须是 子串 的长度，"pwke" 是一个子序列，不是子串。
# 
# 
#
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left, right = 0, 0
        max_len = 0
        memo = {}
        while right < len(s):
            last = memo.get(s[right], None)
            memo[s[right]] = right
            if last is not None:
                max_len = max(max_len, right-left)
                left = max(left, last + 1)
            right += 1
        max_len = max(max_len, right-left)
        return max_len

File: cpp/3.longest-substring-without-repeating-characters.cpp
//  Name: Luke Wang
//  Email: oopswangsl@gmail.com
//  Date: 2023-03-16



//  Given a string s, find the length of the longest substring without repeating characters.
//   
//  Example 1:
//  
//  Input: s = "abcabcbb"
//  Output: 3
//  Explanation: The answer is "abc", with the length of 3.
//  
//  Example 2:
//  
//  Input: s = "bbbbb"
//  Output: 1
//  Explanation: The answer is "b", with the length of 1.
//  
//  Example 3:
//  
//  Input: s = "pwwkew"
//  Output: 3
//  Explanation: The answer is "wke", with the length of 3.
//  Notice that the answer must be a substring, "pwke" is a subsequence and not a substring.
//  
//   
//  Constraints:
//  
//  0 <= s.length <= 5 * 104
//  s consists of English letters, digits, symbols and spaces.
//  

#include <iostream>

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int left = 0, right = 1;
        int max_len = 0, cur_len = 0;
        unordered_set<char> sub_s;

        for (right=0; right<s.length(); right++){
            cur_len++;
            if (sub_s.count(s[right])){
                while (sub_s.count(s[right])){
                    sub_s.erase(s[left]);
                    left++;
                    cur_len--;
                }    
            }
            sub_s.insert(s[right]);
            max_len = max(cur_len, max_len);
        }

        return max_len;
        
    }
};





